### PR TITLE
Record rule usage

### PR DESCRIPTION
**Pull request by QRiary**

### CODE REVIEW

File: keybuffer.c [hunk 1]
```diff
@@ -11,7 +11,7 @@
 #include "keybuffer.h"
 #include "utils.h"
 #include "print.h"
-
+#include "sequence_transform_data.h"
```
Comment (kamih) [inline keybuffer.c]: This file shouldn't need to include the generated data.

File: keybuffer.h [hunk 2]
```diff
@@ -32,3 +33,4 @@ void                    st_key_buffer_reset(st_key_buffer_t *buf);
 void                    st_key_buffer_push(st_key_buffer_t *buf, uint16_t keycode);
 void                    st_key_buffer_pop(st_key_buffer_t *buf, uint8_t num);
 void                    st_key_buffer_print(st_key_buffer_t *buf);
+void                    st_key_buffer_get_context_string(st_key_buffer_t *buf, char* output_string, uint8_t context_len);
```
Comment (kamih) [inline keybuffer.h]: Should be: `st_key_buffer_get_context_string(const st_key_buffer_t *buf, char *output_string, uint8_t context_len);`

File: sequence_transform.c [hunk 3]
```diff
@@ -181,23 +182,65 @@ void st_handle_repeat_key()
     }
 }
 //////////////////////////////////////////////////////////////////////////////////////////
+void log_rule(st_trie_t *trie, st_trie_payload_t *res) {
+    char context_string[SEQUENCE_MAX_LENGTH];
```
Comment (kamih) [inline sequence_transform.c]: Should be `[SEQUENCE_MAX_LENGTH+1]` to leave space for null terminator.

File: sequence_transform.c [hunk 4]
```diff
@@ -181,23 +182,65 @@ void st_handle_repeat_key()
...
+    st_key_buffer_get_context_string(&key_buffer, context_string, res->context_len);
+
+    char rule_trigger_char = context_string[res->context_len - 1]; 
+    context_string[res->context_len - 1] = '\0';
+
+    bool is_repeat = rule_trigger_char == 'R' && strlen(context_string) == 0;
```
Comment (kamih) [inline sequence_transform.c:193]: Don't need to do this now, but should add a TODO comment to not hardcode 'R'.

File: sequence_transform.c [hunk 5]
```diff
@@ -181,23 +182,65 @@ void st_handle_repeat_key()
...
 #ifdef SEQUENCE_TRANSFORM_LOG_GENERAL
     uprintf("completion search res: index: %d, len: %d, bspaces: %d, func: %d\n",
             res->completion_index, res->completion_len, res->num_backspaces, res->func_code);
 #endif
+
 #if defined(RECORD_RULE_USAGE) && defined(CONSOLE_ENABLE)
-    uprintf("used rule: hello world\n");
+    log_rule(trie, res);
 #endif
+
```
Comment (kamih) [inline sequence_transform.c:236]: Please avoid style changes to code that is unrelated to the new features. (Here: these new empty lines.)

File: trie.c [hunk 6]
```diff
@@ -69,12 +70,13 @@ bool st_find_longest_chain(st_trie_t *trie, st_key_buffer_t *search, st_trie_pay
...
             return true;
         // If no better match found deeper, so recordd the payload result!
-        st_get_payload_from_code(res, code, TDATA(offset + 1));
+        st_get_payload_from_code(res, code, TDATA(offset + 1), depth);
 #ifdef SEQUENCE_TRANSFORM_TRIE_SANITY_CHECKS
          // bounds check completion data
         if (res->completion_index + res->completion_len > trie->completions_size) {
             uprintf("find_longest_chain() ERROR: trying to read past end of completion data buffer! index: %d, len: %d\n",
                 res->completion_index, res->completion_len);
+
```
Comment (kamih) [inline trie.c]: Unrelated style change.

**Review by kamih — CHANGES_REQUESTED** [review]
Couple minor things to fix.

File: trie.h [hunk 7]
```diff
@@ -27,12 +27,13 @@ typedef struct
     uint8_t     completion_len;     // length of completion string
     uint8_t     num_backspaces;     // number of backspaces to send before the completion string
     uint8_t     func_code;          // special function code
+    uint8_t     context_len;
```
Comment (Ikcelaks) [inline trie.h]: I believe this should be called `depth` and set equal to the depth that the result was matched in the trie.

You should name a field based on what it IS, not how you plan to use it. The `st_trie_payload` is a struct containing information about a match in the trie. This filed represents how deep into the trie the match was made.

Additionally, calling it `context_len` is even confusing when you use it, because it isn't the length of the context, it's some shorter length equal to the length of the matching portion of the context.

**Review by kamih — APPROVED** [review]
